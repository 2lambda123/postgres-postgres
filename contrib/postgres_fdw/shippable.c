/*-------------------------------------------------------------------------
 *
 * shippable.c
 *	  Facility to track database objects shippable to a foreign server.
 *
 * Determine if functions and operators for non-built-in types/functions/ops
 * are shippable to the remote server.
 *
 * Portions Copyright (c) 1996-2015, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *	  contrib/postgres_fdw/shippable.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "postgres_fdw.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "access/htup_details.h"
#include "catalog/dependency.h"
#include "catalog/indexing.h"
#include "catalog/pg_depend.h"
#include "utils/fmgroids.h"
#include "utils/hsearch.h"
#include "utils/inval.h"
#include "utils/rel.h"
#include "utils/snapmgr.h"
#include "utils/syscache.h"

/* Hash table for informations about remote objects we'll call */
static HTAB *ShippableCacheHash = NULL;

/* objid is the lookup key, must appear first */
typedef struct
{
	/* extension the object appears within, or InvalidOid if none */
	Oid	objid;
	Oid	classid;
} ShippableCacheKey;

typedef struct
{
	/* lookup key - must be first */
	ShippableCacheKey key;
	bool shippable;
} ShippableCacheEntry;

/*
 * Flush all cache entries when pg_foreign_server is updated.
 */
static void
InvalidateShippableCacheCallback(Datum arg, int cacheid, uint32 hashvalue)
{
	HASH_SEQ_STATUS status;
	ShippableCacheEntry *entry;

	hash_seq_init(&status, ShippableCacheHash);
	while ((entry = (ShippableCacheEntry *) hash_seq_search(&status)) != NULL)
	{
		if (hash_search(ShippableCacheHash,
						(void *) &entry->key,
						HASH_REMOVE,
						NULL) == NULL)
			elog(ERROR, "hash table corrupted");
	}
}

/*
 * Initialize the cache of functions we can ship to remote server.
 */
static void
InitializeShippableCache(void)
{
	HASHCTL ctl;

	/* Initialize the hash table. */
	MemSet(&ctl, 0, sizeof(ctl));
	ctl.keysize = sizeof(ShippableCacheKey);
	ctl.entrysize = sizeof(ShippableCacheEntry);
	ShippableCacheHash =
		hash_create("Shippable cache", 256, &ctl, HASH_ELEM);

	CacheRegisterSyscacheCallback(FOREIGNSERVEROID,
								  InvalidateShippableCacheCallback,
								  (Datum) 0);
}

/*
 * Returns true if given operator/function is part of an extension listed in
 * the server options.
 */
static bool
lookup_shippable(Oid objnumber, Oid classnumber, List *extension_list)
{
	static int nkeys = 2;
	ScanKeyData key[nkeys];
	HeapTuple tup;
	Relation depRel;
	SysScanDesc scan;
	bool is_shippable = false;

	/* Always return false if the user hasn't set the "extensions" option */
	if (extension_list == NIL)
		return false;

	depRel = heap_open(DependRelationId, RowExclusiveLock);

	/*
	 * Scan the system dependency table for all entries this object
	 * depends on, then iterate through and see if one of them
	 * is an extension declared by the user in the options
	 */
	ScanKeyInit(&key[0],
				Anum_pg_depend_classid,
				BTEqualStrategyNumber, F_OIDEQ,
				ObjectIdGetDatum(classnumber));

	ScanKeyInit(&key[1],
				Anum_pg_depend_objid,
				BTEqualStrategyNumber, F_OIDEQ,
				ObjectIdGetDatum(objnumber));

	scan = systable_beginscan(depRel, DependDependerIndexId, true,
							  GetCatalogSnapshot(depRel->rd_id), nkeys, key);

	while (HeapTupleIsValid(tup = systable_getnext(scan)))
	{
		Form_pg_depend foundDep = (Form_pg_depend) GETSTRUCT(tup);

		if (foundDep->deptype == DEPENDENCY_EXTENSION &&
			list_member_oid(extension_list, foundDep->refobjid))
		{
			is_shippable = true;
			break;
		}
	}

	systable_endscan(scan);
	relation_close(depRel, RowExclusiveLock);

	return is_shippable;
}

/*
 * is_shippable
 *     Is this object (proc/op/type) shippable to foreign server?
 *     Check cache first, then look-up whether (proc/op/type) is
 *     part of a declared extension if it is not cached.
 */
bool
is_shippable(Oid objnumber, Oid classnumber, List *extension_list)
{
	ShippableCacheKey key;
	ShippableCacheEntry *entry;

	/* Always return false if the user hasn't set the "extensions" option */
	if (extension_list == NIL)
		return false;

	/* Find existing cache, if any. */
	if (!ShippableCacheHash)
		InitializeShippableCache();

	/* Zero out the key */
	memset(&key, 0, sizeof(key));

	key.objid = objnumber;
	key.classid = classnumber;

	entry = (ShippableCacheEntry *)
				 hash_search(ShippableCacheHash,
					(void *) &key,
					HASH_FIND,
					NULL);

	/* Not found in ShippableCacheHash cache.  Construct new entry. */
	if (!entry)
	{
		/*
		 * Right now "shippability" is exclusively a function of whether
		 * the obj (proc/op/type) is in an extension declared by the user.
		 * In the future we could additionally have a whitelist of functions
		 * declared one at a time.
		 */
		bool shippable = lookup_shippable(objnumber, classnumber, extension_list);

		/*
		 * Don't create a new hash entry until *after* we have the shippable
		 * result in hand, as the shippable lookup might trigger a cache
		 * invalidation.
		 */
		entry = (ShippableCacheEntry *)
					 hash_search(ShippableCacheHash,
						(void *) &key,
						HASH_ENTER,
						NULL);

		entry->shippable = shippable;
	}

	if (!entry)
		return false;
	else
		return entry->shippable;
}
