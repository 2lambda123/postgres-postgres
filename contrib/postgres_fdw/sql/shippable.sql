-- ===================================================================
-- create FDW objects
-- ===================================================================

-- Error, extension isn't installed yet
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');

-- Try again
CREATE EXTENSION cube;
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
ALTER SERVER loopback OPTIONS (DROP extensions);


-- ===================================================================
-- create objects used through FDW loopback server
-- ===================================================================

CREATE SCHEMA "SH 1";
CREATE TABLE "SH 1"."TBL 1" (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
);

INSERT INTO "SH 1"."TBL 1"
	SELECT id,
	       2 * id,
	       cube(id,2*id),
	       '1970-01-01'::timestamptz + ((id % 100) || ' days')::interval
	FROM generate_series(1, 1000) id;

ANALYZE "SH 1"."TBL 1";

-- ===================================================================
-- create foreign table
-- ===================================================================

CREATE FOREIGN TABLE shft1 (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
) SERVER loopback
OPTIONS (schema_name 'SH 1', table_name 'TBL 1');

-- ===================================================================
-- simple queries
-- ===================================================================

-- without operator shipping
EXPLAIN (COSTS false) SELECT * FROM shft1 LIMIT 1;
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;

-- with operator shipping
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
EXPLAIN VERBOSE SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;
SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;

-- ===================================================================
-- add a second server with different extension shipping
-- ===================================================================

DO $d$
    BEGIN
        EXECUTE $$CREATE SERVER loopback_two FOREIGN DATA WRAPPER postgres_fdw
            OPTIONS (dbname '$$||current_database()||$$',
                     port '$$||current_setting('port')||$$'
            )$$;
    END;
$d$;

CREATE USER MAPPING FOR CURRENT_USER SERVER loopback_two;

CREATE EXTENSION seg;

CREATE TABLE seg_local (
	id integer,
	s seg,
  n text
);

INSERT INTO seg_local (id, s, n) VALUES (1, '1.0 .. 2.0', 'foo');
INSERT INTO seg_local (id, s, n) VALUES (2, '3.0 .. 4.0', 'bar');
INSERT INTO seg_local (id, s, n) VALUES (3, '5.0 .. 6.0', 'baz');

ANALYZE seg_local;

CREATE FOREIGN TABLE seg_remote_two (
  id integer,
  s seg,
  n text
) SERVER loopback_two
OPTIONS (table_name 'seg_local');

SELECT id FROM seg_local WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
EXPLAIN VERBOSE SELECT id FROM seg_remote_two WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
ALTER SERVER loopback_two OPTIONS (ADD extensions 'seg');
EXPLAIN VERBOSE SELECT id FROM seg_remote_two WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';

CREATE FOREIGN TABLE seg_remote_one (
  id integer,
  s seg,
  n text
) SERVER loopback
OPTIONS (table_name 'seg_local');

SELECT id FROM seg_remote_one WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
EXPLAIN VERBOSE SELECT id FROM seg_remote_one WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
EXPLAIN VERBOSE SELECT id FROM seg_remote_two WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';


-- ===================================================================
-- clean up
-- ===================================================================
DROP FOREIGN TABLE seg_remote_one, seg_remote_two;
DROP USER MAPPING FOR CURRENT_USER SERVER loopback_two;
DROP SERVER loopback_two;
DROP TABLE seg_local;
DROP FOREIGN TABLE shft1;
DROP TABLE "SH 1"."TBL 1";
DROP SCHEMA "SH 1";
DROP EXTENSION cube;
DROP EXTENSION seg;
ALTER SERVER loopback OPTIONS (DROP extensions);
