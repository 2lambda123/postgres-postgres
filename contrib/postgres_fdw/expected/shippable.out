-- ===================================================================
-- create FDW objects
-- ===================================================================
-- Error, extension isn't installed yet
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
ERROR:  required extension "cube" is not installed
HINT:  Extension must be installed locally before it can be used on a remote server.
-- Try again
CREATE EXTENSION cube;
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
ALTER SERVER loopback OPTIONS (DROP extensions);
-- ===================================================================
-- create objects used through FDW loopback server
-- ===================================================================
CREATE SCHEMA "SH 1";
CREATE TABLE "SH 1"."TBL 1" (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
);
INSERT INTO "SH 1"."TBL 1"
	SELECT id,
	       2 * id,
	       cube(id,2*id),
	       '1970-01-01'::timestamptz + ((id % 100) || ' days')::interval
	FROM generate_series(1, 1000) id;
ANALYZE "SH 1"."TBL 1";
-- ===================================================================
-- create foreign table
-- ===================================================================
CREATE FOREIGN TABLE shft1 (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
) SERVER loopback
OPTIONS (schema_name 'SH 1', table_name 'TBL 1');
-- ===================================================================
-- simple queries
-- ===================================================================
-- without operator shipping
EXPLAIN (COSTS false) SELECT * FROM shft1 LIMIT 1;
         QUERY PLAN          
-----------------------------
 Limit
   ->  Foreign Scan on shft1
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
                             QUERY PLAN                              
---------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..205.06 rows=15 width=4)
   Output: c2
   Filter: (shft1.c3 && '(1.5),(2.5)'::cube)
   Remote SQL: SELECT c2, c3 FROM "SH 1"."TBL 1"
(4 rows)

SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
 c2 
----
  2
  4
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..205.06 rows=15 width=4)
   Output: c2
   Filter: (shft1.c3 && '(1.5),(2.5)'::cube)
   Remote SQL: SELECT c2, c3 FROM "SH 1"."TBL 1"
(4 rows)

-- with operator shipping
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..146.86 rows=15 width=4)
   Output: c2
   Remote SQL: SELECT c2 FROM "SH 1"."TBL 1" WHERE ((c3 OPERATOR(public.&&) '(1.5),(2.5)'::public.cube))
(3 rows)

SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
 c2 
----
  2
  4
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..146.86 rows=15 width=4)
   Output: c2
   Remote SQL: SELECT c2 FROM "SH 1"."TBL 1" WHERE ((c3 OPERATOR(public.&&) '(1.5),(2.5)'::public.cube))
(3 rows)

EXPLAIN VERBOSE SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..128.43 rows=7 width=32)
   Output: cube_dim(c3)
   Remote SQL: SELECT c3 FROM "SH 1"."TBL 1" WHERE ((c3 OPERATOR(public.&&) '(1.5),(2.5)'::public.cube))
(3 rows)

SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
 cube_dim 
----------
        1
        1
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;
                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Limit  (cost=100.00..107.22 rows=2 width=4)
   Output: c2
   ->  Foreign Scan on public.shft1  (cost=100.00..154.18 rows=15 width=4)
         Output: c2
         Remote SQL: SELECT c2 FROM "SH 1"."TBL 1" WHERE ((public.cube_dim(c3) = 1))
(5 rows)

SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;
 c2 
----
  2
  4
(2 rows)

-- ===================================================================
-- add a second server with different extension shipping
-- ===================================================================
DO $d$
    BEGIN
        EXECUTE $$CREATE SERVER loopback_two FOREIGN DATA WRAPPER postgres_fdw
            OPTIONS (dbname '$$||current_database()||$$',
                     port '$$||current_setting('port')||$$'
            )$$;
    END;
$d$;
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback_two;
CREATE EXTENSION seg;
CREATE TABLE seg_local (
	id integer,
	s seg,
  n text
);
INSERT INTO seg_local (id, s, n) VALUES (1, '1.0 .. 2.0', 'foo');
INSERT INTO seg_local (id, s, n) VALUES (2, '3.0 .. 4.0', 'bar');
INSERT INTO seg_local (id, s, n) VALUES (3, '5.0 .. 6.0', 'baz');
ANALYZE seg_local;
CREATE FOREIGN TABLE seg_remote_two (
  id integer,
  s seg,
  n text
) SERVER loopback_two
OPTIONS (table_name 'seg_local');
SELECT id FROM seg_local WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
 id 
----
  3
(1 row)

EXPLAIN VERBOSE SELECT id FROM seg_remote_two WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Foreign Scan on public.seg_remote_two  (cost=100.00..157.88 rows=1 width=4)
   Output: id
   Filter: (seg_remote_two.s && '5.8 .. 6.2'::seg)
   Remote SQL: SELECT id, s FROM public.seg_local WHERE ((n = 'baz'::text))
(4 rows)

ALTER SERVER loopback_two OPTIONS (ADD extensions 'seg');
EXPLAIN VERBOSE SELECT id FROM seg_remote_two WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
                                                           QUERY PLAN                                                           
--------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan on public.seg_remote_two  (cost=100.00..153.89 rows=1 width=4)
   Output: id
   Remote SQL: SELECT id FROM public.seg_local WHERE ((s OPERATOR(public.&&) '5.8 .. 6.2'::public.seg)) AND ((n = 'baz'::text))
(3 rows)

CREATE FOREIGN TABLE seg_remote_one (
  id integer,
  s seg,
  n text
) SERVER loopback
OPTIONS (table_name 'seg_local');
SELECT id FROM seg_remote_one WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
 id 
----
  3
(1 row)

EXPLAIN VERBOSE SELECT id FROM seg_remote_one WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Foreign Scan on public.seg_remote_one  (cost=100.00..157.88 rows=1 width=4)
   Output: id
   Filter: (seg_remote_one.s && '5.8 .. 6.2'::seg)
   Remote SQL: SELECT id, s FROM public.seg_local WHERE ((n = 'baz'::text))
(4 rows)

EXPLAIN VERBOSE SELECT id FROM seg_remote_two WHERE s && '5.8 .. 6.2'::seg AND n = 'baz';
                                                           QUERY PLAN                                                           
--------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan on public.seg_remote_two  (cost=100.00..153.89 rows=1 width=4)
   Output: id
   Remote SQL: SELECT id FROM public.seg_local WHERE ((s OPERATOR(public.&&) '5.8 .. 6.2'::public.seg)) AND ((n = 'baz'::text))
(3 rows)

-- ===================================================================
-- clean up
-- ===================================================================
DROP FOREIGN TABLE seg_remote_one, seg_remote_two;
DROP USER MAPPING FOR CURRENT_USER SERVER loopback_two;
DROP SERVER loopback_two;
DROP TABLE seg_local;
DROP FOREIGN TABLE shft1;
DROP TABLE "SH 1"."TBL 1";
DROP SCHEMA "SH 1";
DROP EXTENSION cube;
DROP EXTENSION seg;
ALTER SERVER loopback OPTIONS (DROP extensions);
